ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/commons/objectivefunctionevaluator/MinMarginViolationEvaluator.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) 2024, RTE (http://www.rte-france.com)
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package com.powsybl.openrao.searchtreerao.commons.objectivefunctionevaluator;
+
+import com.powsybl.openrao.commons.Unit;
+import com.powsybl.openrao.data.cracapi.cnec.FlowCnec;
+import com.powsybl.openrao.searchtreerao.commons.marginevaluator.MarginEvaluator;
+import com.powsybl.openrao.searchtreerao.result.api.FlowResult;
+
+import java.util.Set;
+
+/**
+ * @author Thomas Bouquet {@literal <thomas.bouquet at rte-france.com>}
+ */
+public class MinMarginViolationEvaluator extends MinMarginEvaluator implements CostEvaluator {
+    private static final double OVERLOAD_PENALTY = 10000d; // TODO : set this in RAO parameters
+
+    public MinMarginViolationEvaluator(Set<FlowCnec> flowCnecs, Unit unit, MarginEvaluator marginEvaluator) {
+        super(flowCnecs, unit, marginEvaluator);
+    }
+
+    @Override
+    public String getName() {
+        return "min-margin-violation-evaluator";
+    }
+
+    @Override
+    protected double computeCostForState(FlowResult flowResult, Set<FlowCnec> flowCnecsOfState) {
+        return Math.max(0, super.computeCostForState(flowResult, flowCnecsOfState)) * OVERLOAD_PENALTY;
```

Comment:
I think a cost per unit is better because if we use a fixed value, then the RAO has no point using more remedial actions to reduce the overload.

For instance, say that you spent 1000€ in remedial actions and there is still a 15 MW overload. We can still use remedial actions but none can secure the network. However, there's one that costs 100€ and that helps reducing the overload to 2 MW.
If we use a constant cost, the virtual cost for overload will be the same, no matter if we apply the 100€-remedial action so the RAO won't choose it.
On the other hand, using a cost per unit (say 10k / unit), we'll go from a cost of 151000 to 21100 so the RAO will chose to apply the last remedial action to reduce the overload.

It all depends on whether we also try to minimize the overload, what are your thoughts on this?

Reply:
You're right it makes sense to try minimize the excess in the cases where there's not enough linear range actions to solve all overloads for instance.